sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp

```diff
@@ -63,7 +63,7 @@ cdef class ArgKmin{{name_suffix}}(BaseDistancesReduction{{name_suffix}}):
         """
         if (
             metric in ("euclidean", "sqeuclidean")
-            and not (issparse(X) or issparse(Y))
+            and not (issparse(X) ^ issparse(Y))
         ):
```

Comment:
Could you please update the comment below to make it explicit that this optimization is available for for Dense-Dense and Sparse-Sparse inputs but not for the other cases?

I find the condition hard to reason about (negation of a XOR) but a simple comment would do.